Add repositoryMirrors spec to ImageContentSourcePolicy (#690)
openshift/enhancements: enhancements/api-review/add-repositoryMirrors-spec.md

- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)

## Summary

Today, the ImageContentSourcePolicy object sets up mirror with `mirror-by-digest-only` property set to true. Requires images be pulled by digest only. This enhancement plans to add new spec to ImageContentSourcePolicy so mirrors can be used without setting up default `mirror-by-digest-only = true`.

## Motivation

Today, the ImageContentSourcePolicy object sets up mirror configuration with `mirror-by-digest-only` property set to true, which leads to using mirrors only when images are referenced by digests. However, when working with disconnected environments, sometimes multiple ImageContentSourcePolicies are needed, some of them are used by apps/manifests that don't use digests when pulling the images.

Adding the `repositoryMirrors` spec to ImageContentSourcePolicy to configure mirrors without setting `mirror-by-digest-only` will make it possible to pull images from mirror using tags without requiring image digest reference.

### Goals

- Enable the user to pull images through ImageContentSourcePolicy configured mirrors by tags.

### Non-Goals

## Proposal

> The Image API has an optional `MirrorByDigestOnly` filed with type `bool` to `Registry`.

Reviewer: Which Image API? Are we talking `registries.conf` API for containers/image, or the OpenShift ImageStream APIs?

Author: The image API is registries.conf

Reviewer: Please clarify then - there is a separate OpenShift Image API which makes the sentence ambiguous.

```go
// Registry represents a registry.
type Registry struct {
  // ...

	// If true, mirrors will only be used for digest pulls. Pulling images by
	// tag can potentially yield different images, depending on which endpoint
	// we pull from.  Forcing digest-pulls for mirrors avoids that issue.

> 	MirrorByDigestOnly bool `toml:"mirror-by-digest-only,omitempty"`

Reviewer 2: Defaulting to the unsafe `false` is riskier than I'd like.  If this is a true boolean in `registries.conf` (I don't know), I'd prefer `AllowTags` or some such here to default safe.

Reviewer: Isn't mirror by digest the default behavior today? If that is the case then then new API needs to be opt-in, as suggested with `AllowMirrorByTags`

Author: These lines of API definition is current `registries.conf`. Not the ImageCententSourcePolicy that this enhancement try to implement.

Author: > Isn't mirror by digest the default behavior today? If that is the case then then new API needs to be opt-in, as suggested with `AllowMirrorByTags`

I have updated this enhancement to add a new spec of `repositoryMirrros` per comment https://github.com/openshift/enhancements/pull/690#discussion_r632822088

metadata:
  name: ubi8repo
spec:
  repositoryMirrors:
  - mirrors:
    - example.io/example/ubi-minimal 
    source: registry.access.redhat.com/ubi8/ubi-minimal
```

### User Stories

#### As a user, I would like to pull images from mirror using tag, without digests reference
The user need to define multiple ImageContentSourcePolicies, used by apps/manifests that don't use digests to pull
the images when working with disconnected environments or pulling from registries that act as transparent pull-through proxy cache.
The user can pull image without digest by configuring `repositoryMirrors` spec in the ImageContentSourcePolicy file.
And create the ImageContentSourcePolicy project. Once this is done, the images can be pulled from the mirrors without the digest referenced.

### Implementation Details/Notes/Constraints [optional]

Implementing this enhancement requires changes in:
- openshift/api
- openshift/machine-config-operator

This is an example of the ImageContentSourcePolicy file:

```yaml
apiVersion: operator.openshift.io/v1alpha1
kind: ImageContentSourcePolicy
metadata:
  name: ubi8repo
spec:
  repositoryMirrors:
  - mirrors:
    - example.io/example/ubi-minimal 

>     source: registry.access.redhat.com/ubi8/ubi-minimal 
> ```
> 
> The ImageContentSourcePolicy file will create a drop-in file at `/etc/containers/registries.conf` that looks like:
> 
> ```toml
> unqualified-search-registries = ["registry.access.redhat.com", "docker.io"]
> [[registry]]
>   location = "registry.access.redhat.com/ubi8/"
>   insecure = false
>   blocked = false
>   # No mirror-by-digest-only = true configured
>   prefix = ""

Reviewer: Echoing my prior comment about having this behavior be opt-in, ICSP should expose an API to allow opt in:

```suggestion
    source: registry.access.redhat.com/ubi8/ubi-minimal 
    allowMirrorByTag: true
\`\`\`

The ImageContentSourcePolicy file will create a drop-in file at `/etc/containers/registries.conf` that looks like:

\`\`\`toml
unqualified-search-registries = ["registry.access.redhat.com", "docker.io"]
[[registry]]
  location = "registry.access.redhat.com/ubi8/"
  insecure = false
  blocked = false
  allowMirrorByTag = true
  prefix = ""
```

Author: The above is /etc/containers/registries.conf, it already have mirror-by-digest-only. We don't plan to change it in this enhancement.

Author: Fixed by adding allowMirrorByTag as opt-in.

  [[registry.mirror]]
    location = "example.io/example/ubi8-minimal"
    insecure = false
```

> If both `repositoryMirrors` and `repositoryDigestMirrors` are configured for same “source” repository, the ImageContentSourcePolicy will be rejected.

Reviewer 2: With a property for digest-ness under `repositoryMirrors`, maybe time to go v1alpha2 and drop `repositoryDigestMirrors`?

Author: If the repositoryDigestMirrors dropped, the digest requirement will not exist. Is it possible the customer still wants to limit the use of mirrors? This looks like a very rare case, I am not sure.

Reviewer 2: `mirror-by-digest-only: true` is sane, and protects you from malicious/compromised registries by forcing you to use by-digest pullspecs.  But you can ask for one if those with:

```yaml
apiVersion: operator.openshift.io/v1alpha1
kind: ImageContentSourcePolicy
metadata:
  name: ubi8repo
spec:
  repositoryMirrors:
  - mirrors:
    - example.io/example/ubi-minimal 
    source: registry.access.redhat.com/ubi8/ubi-minimal
  allowMirrorByTags: false
```

without needing a `repositoryDigestMirrors` entry.

Author: > With a property for digest-ness under `repositoryMirrors`, maybe time to go v1alpha2 and drop `repositoryDigestMirrors`?

@wking  How we do the API version change and the drop for `repositoryDigestMirrors`? Do we add the new `repositoryMirrors` and drop the old one at the same time in openshift/api?

Should this enhancement mention we are going to drop the `repositoryDigestMirrors`?

### Risks and Mitigations

> To pull images from mirror registries without the digest specifications, this could lead to returning different image version from different registry if the image tag mapping is out of sync. This does not bring risks for ICSP usage since it does not influence the openshift deployments, which required image using digests to avoid different versions.

Reviewer 2: I think the "This does not bring risks..." sentence is a bit off base.  It would be more clear to say the OpenShift core won't consume this feature at all, so it is not exposed to the risks that anyone who actually uses the feature will be exposed to.

- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)

## Summary

Today, the ImageContentSourcePolicy object sets up mirror with `mirror-by-digest-only` property set to true. Requires images be pulled by digest only. This enhancement plans to add new spec to ImageContentSourcePolicy with configurable field `allowMirrorByTags`, so that mirrors can be used without setting up default `mirror-by-digest-only = true`.

## Motivation

Today, the ImageContentSourcePolicy object sets up mirror configuration with `mirror-by-digest-only` property set to true, which leads to using mirrors only when images are referenced by digests. However, when working with disconnected environments, sometimes multiple ImageContentSourcePolicies are needed, some of them are used by apps/manifests that don't use digests when pulling the images.

Adding the `repositoryMirrors` spec to ImageContentSourcePolicy to configure mirrors will make it possible to pull images from mirror using tags without requiring image digest reference.

### Goals

- Enable the user to pull images through ImageContentSourcePolicy configured mirrors by tags.

> ### Non-Goals

Reviewer 3: Add a bullet that we won't enable this setting for OpenShift payload images.

Reviewer 2: done in the current iteration.  Can we mark this thread resolved?

---
title: add-repositoryMirrors-spec
authors:
  - "@QiWang19"
reviewers:
  - TBD
approvers:
  - TBD
creation-date: 2021-03-10
last-updated: 2021-03-10
status: implementable
---

# Add repositoryMirrors spec to cluster wide ImageContentSourcePolicy

## Release Signoff Checklist

- [x] Enhancement is `implementable`
- [x] Design details are appropriately documented from clear requirements
- [x] Test plan is defined
- [ ] Operational readiness criteria is defined
- [ ] Graduation criteria for dev preview, tech preview, GA
- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)

## Summary

> Today, the ImageContentSourcePolicy object sets up mirror with `mirror-by-digest-only` property set to true. Requires images be pulled by digest only. This enhancement plans to add new spec to ImageContentSourcePolicy with configurable field `allowMirrorByTags`, so that mirror-by-digest-only` can be configured by ImageContentSourcePolicy.

Reviewer 2: nit: unbalanced backticks around the second `mirror-by-digest-only`.

## Motivation

Today, the ImageContentSourcePolicy object sets up mirror configuration with `mirror-by-digest-only` property set to true, which leads to using mirrors only when images are referenced by digests. However, when working with disconnected environments, sometimes multiple ImageContentSourcePolicies are needed, some of them are used by apps/manifests that don't use digests when pulling the images.

Adding the `repositoryMirrors` spec to ImageContentSourcePolicy to configure mirrors will make it possible to pull images from mirror using tags without requiring image digest reference.

### Goals

- Enable the user to pull images through ImageContentSourcePolicy configured mirrors by tags.

### Non-Goals

> - The `allowMirrorByTags` will not be configurable or applied to OpenShift payload images.

Reviewer 2: This wording seems to have been inspired by [this comment][1], but I think the wording could be more clear.  Maybe you're aiming for something like:

> - This proposal does not recommend using by-tag references for OpenShift release images.  Those should still be referenced by digest, regardless of whether `allowMirrorByTags` is enabled for repositories where the release images are mirrored.

[1]: https://github.com/openshift/enhancements/pull/690#discussion_r655659846

## Proposal

> The new spec `repositoryMirrors` will be added to ImageContentSourcePolicy. The `allowMirrorByTags` will be the configurable filed can allow using mirrors without requiring digest reference.

Reviewer 2: Maybe the second sentence could be rephrased to:

> The `allowMirrorByTags` property will allow for mirroring by-tag images through configured mirrors.

or some such.

```go
// RepositoryMirrors holds cluster-wide information about how to handle mirrors in the registries config.
// Note: this is different from the RepositoryDigestMirrors that the mirrors only 
// work when pulling the images that are referenced by their digests.
type RepositoryMirrors struct {
	// source is the repository that users refer to, e.g. in image pull specifications.
	// +required
	Source string `json:"source"`

>   // If true, mirrors will only be used for digest pulls. Pulling images by

Reviewer 2: nit: inconsistent indenting flipping between tabs and spaces.

Author: Fixed.

    // tag can potentially yield different images, depending on which endpoint
    // we pull from.  Forcing digest-pulls for mirrors avoids that issue.
    // +optional
    AllowMirrorByTags bool `json:"allowMirrorByTags"`
    // mirrors is one or more repositories that may also contain the same images.
    // The order of mirrors in this list is treated as the user's desired priority, while source

>     // is by default considered lower priority than all mirrors. Other cluster configuration,

Reviewer 4: Specifically "source being last" only makes sense for digest, it DOES not make sense for tag in all use cases.  Tag is fundamentally dependent on hoping you get the most up to date version of a tag during upgrade scenarios.  

Both image stream import and crio can use repository digest order consistently.  However, I'm not sure without an understanding of image import tag mirroring what we do here.

Is there another enhancement that covers ImageStream import behavior with repository tag design?  We should not be introducing a V1 api without a full design covering how import will interpret this, because mirroring of tags may actually need these feels to be designed.  If there is no additional design, I'd like to see one before I withdraw my hold on the implementation.

    // including (but not limited to) other repositoryMirrors objects,
    // may impact the exact order mirrors are contacted in, or some mirrors may be contacted
    // in parallel, so this should be considered a preference rather than a guarantee of ordering.
    // +optional
    Mirrors []string `json:"mirrors"`
}
```

The ImageContentSourcePolicy CRD has the cluster-wide information about handle the mirrors only when pulling the images that are referenced by their digests.
It will now watch for mirrors configured through spec `repositoryMirrors` to not require digest reference.

An example ImageContentSourcePolicy file will look like:
```yaml
apiVersion: operator.openshift.io/v1alpha1
kind: ImageContentSourcePolicy
metadata:
  name: ubi8repo
spec:
  repositoryMirrors:
  - mirrors:
    - example.io/example/ubi-minimal 
    source: registry.access.redhat.com/ubi8/ubi-minimal
    allowMirrorByTags: false # or equivalently, by leaving allowMirrorByTags unset

> ```

Reviewer 4: I would probably recommend defining composition of mirrors in name order (ICSP A mirrors go before ICSP B mirrors)

## Design Details

### Test Plan

Update the tests that are currently in the MCO to verify that `mirror-by-digest-only` not have been set when spec repositoryMirrors of the ImageContentSourcePolicy is created.

### Graduation Criteria

GA in openshift will be looked at to
determine graduation.

#### Dev Preview -> Tech Preview

- Ability to utilize the enhancement end to end
- End user documentation, relative API stability
- Sufficient test coverage
- Gather feedback from users rather than just developers
- Enumerate service level indicators (SLIs), expose SLIs as metrics
- Write symptoms-based alerts for the component(s)

#### Tech Preview -> GA

- More testing (upgrade, downgrade, scale)
- Sufficient time for feedback
- Available by default
- Backhaul SLI telemetry
- Document SLOs for the component
- Conduct load testing

#### Removing a deprecated feature

> - With `repositoryMirrors` getting in, the current `repositoryDigestMirrors` can be deprecated since its functionality will also be fully satisfied by `repositoryMirrors` and we don't have to keep duplicated APIs.

Reviewer 4: This isn't a very strong argument, changing behavior without understanding the difference in behavior is a bug.

So I would probably put a hold on this feature with this as is without stronger justification.
